snuba/datasets/entity.py

```diff
@@ -1,27 +1,16 @@
 from abc import ABC, abstractmethod
-from typing import Mapping, NamedTuple, Optional, Sequence
+from typing import Mapping, Optional, Sequence
 
 from snuba.clickhouse.columns import ColumnSet
-from snuba.datasets.entities import EntityKey
 from snuba.datasets.plans.query_plan import ClickhouseQueryPlan
 from snuba.datasets.storage import Storage, WritableStorage, WritableTableStorage
 from snuba.pipeline.query_pipeline import QueryPipelineBuilder
-from snuba.query.data_source.join import JoinClass, JoinCondition
+from snuba.query.data_source.join import JoinRelationship
 from snuba.query.extensions import QueryExtension
 from snuba.query.processors import QueryProcessor
 from snuba.query.validation import FunctionCallValidator
 
 
-class JoinRelationship(NamedTuple):
```

Comment:
I moved it to the join module because it made more sense to be there (otherwise I am importing the entity module all over the snql code for this one data structure). While I understand conceptually why we have join class, it seemed like unnecessary work to look up the join type and join modifier from the class. So I added them directly to the relationship.

Reply:
Doesn't the join class imply a join type and join modifier? I thought that a join class was essentially a renaming of the join type/modifier combination. I started writing it that way and then it seemed like a lot of boilerplate to convert from `JoinClass.ONE_2_ONE` to `(JoinType.INNER, JoinModifier.SEMI)` or something equivalent.

Also, the `JoinRelationship` class is not being used for two purposes. The relationship is only read from the Entity, it's never modified. I don't know why you think there needs to be two different classes.

Reply:
I think we discussed this topic here: 
https://github.com/getsentry/snuba/pull/1484#discussion_r521729274

The join modifier (SEMI) Is not a property of the relationship. It is a property of the query. It can be applied to any relationship (1/1, 1/n, ....). It is something we will apply when we can during join optimization to avoid a full cartesian product (that code is not yet there). 
Whether we can do that depends on the full parsed query. Basically we can instruct clickhouse to do a SEMI join if the user does not want any column from the right side table and it is only used for filtering, or if the data on the right side is aggregated.
So it is orthogonal with respect to the join type or join class and it is applied on the fly.

The join class can imply a join type (not the modifier). From the discussion we had on the previous PR it seemed to me we agreed that, the join class at a logical level represented the data model over a graph more appropriately since it preserve information about the cardinality (1 vs n) that the join class does not represent.
This is not a major concern to me, so if you do not agree with that we can use JoinType instead of JoinClass as we do not really use that cardinality information yet.
Though in such case please remove JoinClass entirely and still the modifier has no place there. So two structures still.

You can reduce the boilerplate by adding a `fromRelationship` method in the class you would use in the parser that builds it from a JoinRelationship object or by making the class you need in the parser extend JoinRelationship

Reply:
OK I removed the join modifier from JoinRelationship. I kept JoinType and JoinClass. Since these now map 1:1 it's not clear to me why we need JoinClass, but I left it in since I think there's further discussion to be had around this and can be dealt with in a separate PR.

Reply:
> it's not clear to me why we need JoinClass

We don't if we add JoinType here. We loose information around the structure of the logical data model that could be used to optimize the query (knowing that the data will be 1:n instead of 1:1 can drive us to use SEMI joins in some additional situations) but we do not use it now and we can add it back later if it becomes really useful.